runtime/src/tiered_storage/file.rs

```diff
@@ -0,0 +1,73 @@
+use std::{
+    fs::{File, OpenOptions},
+    io::{Read, Seek, SeekFrom, Write},
+    mem,
+    path::Path,
+};
+
+#[derive(Debug)]
+pub struct TieredStorageFile {
+    pub file: File,
+}
+
+impl TieredStorageFile {
+    /// Creates a tiered-storage file.
+    /// If the create flag is false, it will open an existing file
+    /// in read-only mode.
+    pub fn new<P: AsRef<Path>>(file_path: P, create: bool) -> Self {
+        let file = OpenOptions::new()
+            .read(true)
+            .write(create)
+            .create(create)
+            .open(file_path.as_ref())
+            .map_err(|e| {
+                panic!(
+                    "Unable to {} data file {} in current dir({:?}): {:?}",
+                    if create { "create" } else { "open" },
+                    file_path.as_ref().display(),
+                    std::env::current_dir(),
+                    e
+                );
+            })
+            .unwrap();
+        Self { file }
+    }
+
+    pub fn write_type<T>(&self, value: &T) -> Result<usize, std::io::Error> {
+        unsafe {
+            let ptr =
+                std::slice::from_raw_parts((value as *const T) as *const u8, mem::size_of::<T>());
+            (&self.file).write_all(ptr)?;
+        }
```

Comment:
> [2]: How does this currently work if self is not &mut? Is it only because we're within unsafe? If yes, that seems a little scary...

It's not because we're within unsafe.  It is because `(&self.file).write_all()` does the magic :).